test/e2e_kubeadm/networking_test.go

```diff
@@ -0,0 +1,151 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_kubeadm
+
+import (
+	"net"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
+
+	. "github.com/onsi/ginkgo"
+	// . "github.com/onsi/gomega"
+)
+
+var (
+	dualStack                    bool
+	podSubnetInKubeadmConfig     bool
+	serviceSubnetInKubeadmConfig bool
+)
+
+// Define container for all the test specification aimed at verifying
+// that kubeadm configures the networking as expected.
+// in case you want to skip this test use SKIP=setup-networking
+var _ = KubeadmDescribe("networking [setup-networking]", func() {
+
+	// Get an instance of the k8s test framework
+	f := framework.NewDefaultFramework("networking")
+
+	// Tests in this container are not expected to create new objects in the cluster
+	// so we are disabling the creation of a namespace in order to get a faster execution
+	f.SkipNamespaceCreation = true
+
+	// Tests are either single-stack (either IPv4 or IPv6 address)	or dual-stack
+	// We can determine this from the kubeadm config by looking at the number of
+	// IPs specified in networking.podSubnet
+	BeforeEach(func() {
+
+		// gets the ClusterConfiguration from the kubeadm kubeadm-config ConfigMap as a untyped map
+		cc := getClusterConfiguration(f.ClientSet)
+		// Extract the networking.podSubnet
+		if _, ok := cc["networking"]; ok {
+			d := cc["networking"].(map[interface{}]interface{})
+			if ps, ok := d["podSubnet"]; ok {
+				// If podSubnet is not specified, podSubnet cases will be skipped.
+				podSubnetInKubeadmConfig = true
+				cidrs := strings.Split(ps.(string), ",")
+				if len(cidrs) > 1 {
+					dualStack = true
+				}
+			}
+			if _, ok := d["serviceSubnet"]; ok {
+				// If serviceSubnet is not specified, serviceSubnet cases will be skipped.
+				serviceSubnetInKubeadmConfig = true
+			}
+		}
+	})
+
+	Context("single-stack", func() {
+		Context("podSubnet", func() {
+			It("should be properly configured if specified in kubeadm-config", func() {
+				if dualStack {
+					framework.Skipf("Skipping because cluster is dual-stack")
+				}
+				if !podSubnetInKubeadmConfig {
+					framework.Skipf("Skipping because podSubnet was not specified in kubeadm-config")
+				}
+				cc := getClusterConfiguration(f.ClientSet)
+				if _, ok := cc["networking"]; ok {
+					d := cc["networking"].(map[interface{}]interface{})
+					if ps, ok := d["podSubnet"]; ok {
+						// Check that the pod CIDR allocated to the node(s) is within the kubeadm-config podCIDR.
+						nodes, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
```

Comment:
This test is checking kubeadm behaviour indirectly, and I'm wondering if instead we should test the kubeadm bit more specifically

Kubeadm only responsibility for networking is to pass the settings to the control plane components (in this case, to translate `podSubnet` into `cluster-cidr`, `allocate-node-cidrs` and `node-cidr-mask-size` flags for the controller manager)

Reply:
What do you have in mind? 
I followed this approach based on what I saw in other tests.

Reply:
Sorry for not being clear.
I think we should have a test that checks kubeadm assigns the expected value to cluster-cidr, allocate-node-cidrs and node-cidr-mask-size (in addition or in replacement to the current test)

Reply:
@fabriziopandini 
`allocate-node-cidrs` is internal to controller-manager and is not exposed, so I don't see any way to verify that. 
`node-cidr-mask-size` is calcuated by kubeadm (see func calcNodeCidrSize). There are some issues with verifying it: 
i. This func currently assumes IPv4 (assigns /24 to IPv6 -- a known problem). 
2. We'll need to change this func to support dual-stack, so  `node-cidr-mask-size` will actually become a slice. 
3. `node-cidr-mask-size` is not written to kubeadm config, so we'll have to duplicate this func in the e2e tests in order to verify it. 
4. calcNodeCidrSize is unit tested already. 
Because of these reasons, I would prefer not to add e2e tests for this. Perhaps when dual-stack changes land, we can revisit this.

Reply:
Ok, I got your point and I don't want to block on this.
However, I think this test is useful and doable reading the args in the controller-manager pod, so please add an item on the umbrella issue: "implement a test that checks kubeadm assigns the expected value to cluster-cidr, allocate-node-cidrs and node-cidr-mask-size (in addition or in replacement to the current test"